ASSERTION FAILED: !m_listeners.contains(connection) || m_listeners.get(connection) == identifier in StorageAreaBase

szewai opened the pull request:
#### dcf41af3936b535cd2875e868290874ab7ac3a65
<pre>
ASSERTION FAILED: !m_listeners.contains(connection) || m_listeners.get(connection) == identifier in StorageAreaBase
<a href="https://bugs.webkit.org/show_bug.cgi?id=251612">https://bugs.webkit.org/show_bug.cgi?id=251612</a>
rdar://104609918

Reviewed by Alex Christensen.

The assertion ensures that web process will not have two StorageAreaMaps that connect to the same StorageArea. We avoid
having two StorageAreaMaps mapping to the same area because StorageAreaMap may need to send sync message for connection
(StorageAreaMap::connectSync()), and we don&apos;t want sync message if unnecessary.

In current implementation, each web process is associated with a session. In web process, each page points to a
WebStorageNameSpaceProvider picked based on page group (WebStorageNamespaceProvider::getOrCreate(WebPageGroupProxy&amp;)).
Each WebStorageNamespaceProvider has one local StorageNamespace (WebCore::StorageNamespaceProvider), and each
StorageNamespace has multiple StorageAreaMaps (keyed by origin). Therefore, in web process, StorageAreaMap is identified
by { SessionID, PageGroupID, ClientOrigin }. However, in network process, StorageArea is identified by
{ SessionID, ClientOrigin }.

We used to have PageGroupID in key because we wanted to separate storage between groups in WebKitLegacy. As we moved
to WebKit, we still have it in key on the client side (web process), but not in the server side (network process), so
two StorageAreaMaps in different page groups can point to the same database, making it pointless to keep PageGroupID.
Therefore, this patch just drops PageGroup from WebStorage in WebKit.

* Source/WebCore/storage/StorageEventDispatcher.cpp:
(WebCore::StorageEventDispatcher::dispatchLocalStorageEvents):
* Source/WebCore/storage/StorageEventDispatcher.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::connectToStorageArea):
(WebKit::NetworkStorageManager::connectToStorageAreaSync):
(WebKit::NetworkStorageManager::cancelConnectToStorageArea):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
(WebKit::WebPage::~WebPage):
* Source/WebKit/WebProcess/WebStorage/StorageAreaMap.cpp:
(WebKit::StorageAreaMap::dispatchLocalStorageEvent):
* Source/WebKit/WebProcess/WebStorage/StorageNamespaceImpl.cpp:
(WebKit::StorageNamespaceImpl::createSessionStorageNamespace):
(WebKit::StorageNamespaceImpl::createLocalStorageNamespace):
(WebKit::StorageNamespaceImpl::createTransientLocalStorageNamespace):
(WebKit::StorageNamespaceImpl::StorageNamespaceImpl):
(WebKit::StorageNamespaceImpl::copy):
(WebKit::StorageNamespaceImpl::pageGroupID const): Deleted.
* Source/WebKit/WebProcess/WebStorage/StorageNamespaceImpl.h:
* Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.cpp:
(WebKit::existingStorageNameSpaceProvider):
(WebKit::WebStorageNamespaceProvider::getOrCreate):
(WebKit::WebStorageNamespaceProvider::incrementUseCount):
(WebKit::WebStorageNamespaceProvider::decrementUseCount):
(WebKit::WebStorageNamespaceProvider::WebStorageNamespaceProvider):
(WebKit::WebStorageNamespaceProvider::createLocalStorageNamespace):
(WebKit::WebStorageNamespaceProvider::createTransientLocalStorageNamespace):
(): Deleted.
(WebKit::WebStorageNamespaceProvider::~WebStorageNamespaceProvider): Deleted.
* Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.h:
* Source/WebKitLegacy/Storage/StorageAreaImpl.cpp:
(WebKit::StorageAreaImpl::dispatchStorageEvent):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LocalStoragePersistence.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/259876@main">https://commits.webkit.org/259876@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/759f9c3f8d0919c5dcb77f1998d8deab4125064c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106273 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15327 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39112 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115466 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/175571 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110183 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16769 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6543 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98489 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115151 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112037 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12757 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95740 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40309 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94631 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27383 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81999 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8567 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28735 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9077 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5302 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14690 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48281 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6825 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10613 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 reviewed (approved):
Looks good, but please write a test that uses WKWebViewConfiguration._groupIdentifier because we're about to remove WKPageGroupCreateWithIdentifier.

szewai committed:
ASSERTION FAILED: !m_listeners.contains(connection) || m_listeners.get(connection) == identifier in StorageAreaBase
https://bugs.webkit.org/show_bug.cgi?id=251612
rdar://104609918

Reviewed by Alex Christensen.

The assertion ensures that web process will not have two StorageAreaMaps that connect to the same StorageArea. We avoid
having two StorageAreaMaps mapping to the same area because StorageAreaMap may need to send sync message for connection
(StorageAreaMap::connectSync()), and we don't want sync message if unnecessary.

In current implementation, each web process is associated with a session. In web process, each page points to a
WebStorageNameSpaceProvider picked based on page group (WebStorageNamespaceProvider::getOrCreate(WebPageGroupProxy&)).
Each WebStorageNamespaceProvider has one local StorageNamespace (WebCore::StorageNamespaceProvider), and each
StorageNamespace has multiple StorageAreaMaps (keyed by origin). Therefore, in web process, StorageAreaMap is identified
by { SessionID, PageGroupID, ClientOrigin }. However, in network process, StorageArea is identified by
{ SessionID, ClientOrigin }.

We used to have PageGroupID in key because we wanted to separate storage between groups in WebKitLegacy. As we moved
to WebKit, we still have it in key on the client side (web process), but not in the server side (network process), so
two StorageAreaMaps in different page groups can point to the same database, making it pointless to keep PageGroupID.
Therefore, this patch just drops PageGroup from WebStorage in WebKit.

* Source/WebCore/storage/StorageEventDispatcher.cpp:
(WebCore::StorageEventDispatcher::dispatchLocalStorageEvents):
* Source/WebCore/storage/StorageEventDispatcher.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::connectToStorageArea):
(WebKit::NetworkStorageManager::connectToStorageAreaSync):
(WebKit::NetworkStorageManager::cancelConnectToStorageArea):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
(WebKit::WebPage::~WebPage):
* Source/WebKit/WebProcess/WebStorage/StorageAreaMap.cpp:
(WebKit::StorageAreaMap::dispatchLocalStorageEvent):
* Source/WebKit/WebProcess/WebStorage/StorageNamespaceImpl.cpp:
(WebKit::StorageNamespaceImpl::createSessionStorageNamespace):
(WebKit::StorageNamespaceImpl::createLocalStorageNamespace):
(WebKit::StorageNamespaceImpl::createTransientLocalStorageNamespace):
(WebKit::StorageNamespaceImpl::StorageNamespaceImpl):
(WebKit::StorageNamespaceImpl::copy):
(WebKit::StorageNamespaceImpl::pageGroupID const): Deleted.
* Source/WebKit/WebProcess/WebStorage/StorageNamespaceImpl.h:
* Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.cpp:
(WebKit::existingStorageNameSpaceProvider):
(WebKit::WebStorageNamespaceProvider::getOrCreate):
(WebKit::WebStorageNamespaceProvider::incrementUseCount):
(WebKit::WebStorageNamespaceProvider::decrementUseCount):
(WebKit::WebStorageNamespaceProvider::WebStorageNamespaceProvider):
(WebKit::WebStorageNamespaceProvider::createLocalStorageNamespace):
(WebKit::WebStorageNamespaceProvider::createTransientLocalStorageNamespace):
(): Deleted.
(WebKit::WebStorageNamespaceProvider::~WebStorageNamespaceProvider): Deleted.
* Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.h:
* Source/WebKitLegacy/Storage/StorageAreaImpl.cpp:
(WebKit::StorageAreaImpl::dispatchStorageEvent):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LocalStoragePersistence.mm:
(TEST):

Canonical link: https://commits.webkit.org/259876@main
